### PR TITLE
Create roc.py

### PR DESCRIPTION
**Pull request by cooperlab**
Functions for calculating time-dependent-roc.

Pull request for line comments.

### CODE REVIEW

**Comment by cooperlab** [comment]
@cramraj8 please add Safoora (safooray) so that I can tag her in comments.

**Comment by cooperlab** [comment]
This provides some of the basic functionality of the R library risksetROC. Implementation mimics the R code for now, but can definitely be streamlined. The functions that were implemented were validated using the 'VA' dataset (I will send to you shortly). The functions produce almost identical results, with the exception of cox_weights, due to some odd numpy behavior.

I will add specific line comments. Please review and also consider how to improve the organization  of functions and interface.

If this is useful we can consider implementing some of the other estimation methods later (local Cox, Schoenfeld).

File: statistics/roc.py [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+    order = np.argsort(at_risk_marker)
+    
+    #calculate ROC curve true positive rates
+    p = np.exp(at_risk_marker) / np.sum(np.exp(at_risk_marker))
+    tp = 1-np.cumsum(p[order])
+    tp = np.insert(tp, 0, 1)
+    tp = np.append(tp, 0)
+    
+    #calculate ROC curve false positive rates
+    p = np.ones(at_risk_marker.shape, dtype=np.float)
```
Comment (cooperlab) [inline statistics/roc.py]: Calculation of 'p' here does not produce exact results. The math is clearly simple but numpy is not able to normalize these to sum to 1 (very odd).

**Comment by cramraj8** [comment]
I sent requests multiple times, but not been acknowledged. I will remind her.

File: statistics/roc.py [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+
+    Parameters
+    ----------
+    aucs : array_like
+        Area under ROC curves at 'times'.    
+    times : array_like
+        Ordered event times (unique).
+    f_surv : array_like
+        Estimated survival function probabilities at 'times'.
+    tmax : float
```
Comment (cooperlab) [inline statistics/roc.py]: I did not test thoroughly with values for tmax that would exclude some samples.

File: statistics/roc.py [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+    
+    #get survival function probability for time just beyond tmax
+    f_surv_max = f_surv[min(i_max, f_surv.size-1)]
+    
+    #calculate weighting
+    if weight == "rescale":
+        w = 2 * f_t * f_surv
+        w = w / w[0:i_max+1].sum()
+    elif weight == "conditional":
+        w = 2 * f_t * (f_surv-f_surv_max) / ((1-f_surv_max)**2)
```
Comment (cooperlab) [inline statistics/roc.py]: Please test this option (weight = "rescale" was tested).

File: statistics/roc.py [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+    
+    #calculate unique event times and survival function
+    times, f_surv = weighted_km(survival, censored, entry=entry)
+    
+    #allocate aucs
+    aucs = np.zeros((times.size,), dtype=np.float)
+    
+    #for each unique time in 'survival', calculate the AUC
+    for i in np.arange(0, times.size):
+        aucs[i], _, _, _ = cox_weights(-0.03342375  * marker, survival, censored,
```
Comment (cooperlab) [inline statistics/roc.py]: Note: the model coefficients produced by the R cox fitter and the lifelines fitter differ significantly (lifelines normalizes feature values internally). This had a significant impact on the AUC values. Please investigate.

File: statistics/roc.py [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+                                        times[i], entry=entry)
+        
+    #calculate c-index
+    c_index = integrate_auc(aucs=aucs, times=times, f_surv=f_surv, 
+                            tmax=tmax, weight=weight)
+    
+    return times, f_surv, aucs, c_index
```
Comment (cooperlab) [inline statistics/roc.py]: Order of outputs seems odd - should re-arrange to put important values first.

Comment (cooperlab) [reply statistics/roc.py]: Check other functions too.

File: statistics/roc.py [hunk 6]
```diff
@@ -0,0 +1,263 @@
+import lifelines as ll
+import numpy as np
+import pandas as pn
+
+def cox_weights(marker, survival, censored, time, entry=None):
```
Comment (cooperlab) [inline statistics/roc.py]: Variable names throughout were chosen for similarity to R code. You may want to change these to maintain consistency with your conventions.

File: statistics/roc.py [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+    ----------
+    marker : array_like
+        Array of predicted risks.
+    survival : array_like
+        Event times.
+    censored : array_like
+        Binary indicator {0, 1} where 1 indicates right-censored samples.
+    time : float
+        Time to calculate ROC curve at.
+    entry : array_like
```
Comment (cooperlab) [inline statistics/roc.py]: I don't know if we ever need to support entry times. I think it's fine to leave it up to users to convert all times to durations (e.g. diagnosis always t=0). You can simplify considerably if you remove this.

File: statistics/roc.py [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+    
+    #calculate unique event times and survival function
+    times, f_surv = weighted_km(survival, censored, entry=entry)
+    
+    #allocate aucs
+    aucs = np.zeros((times.size,), dtype=np.float)
+    
+    #for each unique time in 'survival', calculate the AUC
+    for i in np.arange(0, times.size):
+        aucs[i], _, _, _ = cox_weights(beta * marker, survival, censored,
```
Comment (cooperlab) [inline statistics/roc.py]: Coefficients in R coxph models are significantly different from lifelines (lifelines performs z-score of features internally). Investigate impact on calculating ROC.

File: statistics/roc.py [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+    
+    #binary indicator of events at 'time' that are at_risk
+    events = (survival == time) & (censored == 0)
+    events = events[at_risk]
+    
+    #determine marker sorting order of at risk set
+    order = np.argsort(at_risk_marker)
+    
+    #calculate ROC curve true positive rates
+    p = np.exp(at_risk_marker) / np.sum(np.exp(at_risk_marker))
```
Comment (cooperlab) [inline statistics/roc.py]: This calculation will be significantly impacted by the scale of 'marker' due to the exp. Overall the method will be sensitive to normalization.

File: statistics/roc.py [hunk 10]
```diff
@@ -0,0 +1,263 @@
...
+
+    #binary indicator of at-risk samples
+    at_risk = (survival >= time) & (entry <= time)
+    
+    #marker values of at_risk samples
+    at_risk_marker = marker[at_risk]
+    
+    #binary indicator of events at 'time' that are at_risk
+    events = (survival == time) & (censored == 0)
+    events = events[at_risk]
```
Comment (safooray) [inline statistics/roc.py]: Why is this necessary? at_risk must be a superset of (survival == time) & (censored == 0), right?

File: statistics/roc.py [hunk 11]
```diff
@@ -0,0 +1,263 @@
...
+import pandas as pn
+
+def cox_weights(marker, survival, censored, time, entry=None):
+    """Proportional hazards estimation of true-positive, false-positive, and 
+    area-under-curve for the prediction 'marker' at moment 'time' for the 
+    incident/dynamic ROC curve.
+
+    Parameters
+    ----------
+    marker : array_like
```
Comment (safooray) [inline statistics/roc.py]: Working on renaming some variables. Will make it a pull request that you can later merge with this branch if you approve.

File: statistics/roc.py [hunk 12]
```diff
@@ -0,0 +1,263 @@
...
+        tmax = survival[censored==0].max()
+    
+    #transform data to lifelines duration DataFrame format
+    df = pn.DataFrame(data={'marker': marker, 'survival': survival - entry,
+                            'event': 1-censored})
+    
+    #generate Cox model from marker
+    cph = ll.CoxPHFitter()
+    cph.fit(df, 'survival', event_col='event')
+    beta = cph.hazards_.as_matrix().squeeze()
```
Comment (cooperlab) [inline statistics/roc.py]: Might be better to change this from \beta to \gamma for consistency with paper notation

Comment (safooray) [reply statistics/roc.py]: It's only Gamma because they are fitting a time-dependent Cox regression over the already fit markers. Originally, the coefficients from the PH model are called Beta in the paper.
